conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/ForkDetector.scala

```diff
@@ -0,0 +1,117 @@
+package tech.cryptonomic.conseil.indexer
+
+import tech.cryptonomic.conseil.indexer.ForkDetector._
+
+import cats._
+import cats.implicits._
+import com.typesafe.scalalogging.LazyLogging
+
+/** Definitions and common utilities for fork-detection */
+object ForkDetector {
+
+  /** A data type that express the of search results of a comparison by level */
+  sealed trait IdMatchResult
+
+  case object SameId extends IdMatchResult
+  case object ForkedId extends IdMatchResult
+
+  /** Provides a block identifier, given a blockchain level,
+    * i.e. how high in the chain is the block
+    *
+    * Generically provides a way to verify some chain position to verify
+    * that a forking have happened, by comparing the block between a remote
+    * and local version.
+    *
+    * - BlockId is the type identifying the block. We can't assume anything
+    *   specific here, because each blockchain may have its own type to
+    *   identify blocks uniquely
+    * - Eff is a wrapping effect we expect from the operation call
+    *   we expect that most implementation will require something like Future/IO/Try
+    *   to be returned with the value, because such operation will need to reach
+    *   across the network or on a local service/db in some way
+    * - the search is expected to be able to match the chain internal representation for his
+    *   level/height to an individual [[Long]] value, even if it's actually represented
+    *   via a different type (e.g. a custom typed wrapped like BlockHeight(Int))
+    */
+  trait SearchBlockId[Eff[_], BlockId] {
+    def searchForLevel(level: Long): Eff[BlockId]
+  }
+
+}
+
+/** Defines the algorithms to detect and identify a forking
+  * in the blockchain in comparison to the locally indexed data
+  *
+  * @param indexerSearch run the search on the local indexer
+  * @param nodeSearch run the search on the remote blockchain node
+  * @param monad a [[Monad]] provides ways to combine results in [[Eff]]
```

Comment:
For my understanding can you point out the line where a monad property is used?

Reply:
It's both used when doing `flatMap` call in the `searchForkLevel` at line 108 [might change after updates].

It's where we call a check for comparison of a level, and the result is wrapped by Eff, so we flatMap to compare the results and decide if the bisection should continue and in which binary segment.

Apart from that, we use the `Applicative Functor` power given by a `Monad` to call the `searchForLevel` over both indexer and node in parallel with `mapN`, inside the `checkOnLevel` call.

Reply:
If so requested I could write this down as comments in the code, but I expect that approach to get unwieldy in a short time-lapse